llvm/lib/Target/AMDGPU/AMDGPUNextUseAnalysis.cpp

```diff
@@ -0,0 +1,573 @@
+void AMDGPUNextUseAnalysis::dumpShortestPaths() const {
+  for (const auto &P : ShortestPathTable) {
+    MachineBasicBlock *From = P.first.first;
+    MachineBasicBlock *To = P.first.second;
+    auto [ShortestPath, Dist] = P.second;
+    errs() << "From: " << From->getName() << "-> To:" << To->getName() << " = "
+           << Dist << "\n";
+  }
+}
+
+bool isVCCReg(Register Reg) {
+  return Reg == AMDGPU::VCC || Reg == AMDGPU::VCC_LO || Reg == AMDGPU::VCC_HI;
+}
+
+void AMDGPUNextUseAnalysis::printAllDistances(MachineFunction &MF) {
+  for (MachineBasicBlock &MBB : MF) {
+    for (MachineInstr &MI : *&MBB) {
+      for (const MachineOperand &MO : MI.operands()) {
+        if (!MO.isReg())
+          continue;
+
+        Register Reg = MO.getReg();
+        if (Reg == 0)
+          continue;
+
+        if (MO.isUse())
+          continue;
+
+        if (Reg.isPhysical() || TRI->isAGPR(*MRI, Reg) || isVCCReg(Reg))
+          continue;
+
+        std::optional<uint64_t> NextUseDistance = getNextUseDistance(Reg);
+        errs() << "Next-use distance of Register " << printReg(Reg, TRI)
+               << " = ";
+        if (NextUseDistance)
+          errs() << *NextUseDistance;
+        else
+          errs() << "null";
+        errs() << "\n";
+      }
+    }
+  }
+}
+
+std::optional<uint64_t>
+AMDGPUNextUseAnalysis::getNextUseDistance(Register DefReg) {
+  assert(!DefReg.isPhysical() && !TRI->isAGPR(*MRI, DefReg) &&
+         !isVCCReg(DefReg) &&
+         "Next-use distance is calculated for SGPRs and VGPRs");
+  uint64_t NextUseDistance = std::numeric_limits<uint64_t>::max();
+  uint64_t CurrentNextUseDistance = std::numeric_limits<uint64_t>::max();
+  for (auto &UseMI : MRI->use_nodbg_instructions(DefReg)) {
+    MachineInstr *CurMI = &*MRI->def_instr_begin(DefReg);
+    MachineBasicBlock *CurMBB = CurMI->getParent();
+    MachineBasicBlock *UseMBB = UseMI.getParent();
+    MachineLoop *CurLoop = MLI->getLoopFor(CurMBB);
+    MachineLoop *UseLoop = MLI->getLoopFor(UseMBB);
+
+    bool IsUseOutsideOfTheDefintionLoop =
+        (CurLoop && !UseLoop) ||
+        (CurLoop && UseLoop &&
+         ((!UseLoop->contains(CurLoop) && !CurLoop->contains(UseLoop)) ||
+          (UseLoop->contains(CurLoop) && (UseLoop != CurLoop))));
+
+    if (IsUseOutsideOfTheDefintionLoop) {
+      CurrentNextUseDistance = calculateCurLoopDistance(DefReg, CurMI, &UseMI);
+    } else if (isIncomingValFromBackedge(CurMI, &UseMI, DefReg)) {
+      CurrentNextUseDistance = calculateBackedgeDistance(CurMI, &UseMI);
+    } else {
+      CurrentNextUseDistance = calculateShortestDistance(CurMI, &UseMI);
+    }
+
+    if (CurrentNextUseDistance < NextUseDistance)
+      NextUseDistance = CurrentNextUseDistance;
+  }
+  return NextUseDistance != std::numeric_limits<uint64_t>::max()
+             ? std::optional<uint64_t>(NextUseDistance)
+             : std::nullopt;
+}
+
+std::optional<uint64_t>
+AMDGPUNextUseAnalysis::getNextUseDistance(Register DefReg, MachineInstr *CurMI,
```

Comment:
Now that we have `LiveIntervals` in the `ERS`. You can pass in here, which can help you calculate the distance.